File: SafeVote/migrate_depts.js
const mongoose = require('mongoose');

const MONGO_URI = "mongodb+srv://stharunkumar069_db_user:siet%40123@cluster0.frcnaxx.mongodb.net/safevote?retryWrites=true&w=majority&appName=Cluster0";

const StudentSchema = new mongoose.Schema({
    regNo: { type: Number, required: true, unique: true },
    name: String,
    department: { type: String, default: 'CYBER SECURITY' },
    password: { type: String, default: 'atkboss' },
    hasVoted: { type: Boolean, default: false },
    addedAt: { type: Date, default: Date.now }
});

const Student = mongoose.model('Student', StudentSchema);

async function migrate() {
    try {
        await mongoose.connect(MONGO_URI);
        console.log("Connected to MongoDB for migration...");

        const students = await Student.find({});
        console.log(`Found ${students.length} students to check.`);

        let updatedCount = 0;
        for (const student of students) {
            const sId = student.regNo.toString();
            const deptCode = sId.substring(6, 9);

            let dept = "CYBER SECURITY";
            if (deptCode === "107") {
                dept = "CYBER SECURITY";
            } else if (deptCode === "202") {
                dept = "AIML";
            }

            if (student.department !== dept) {
                student.department = dept;
                await student.save();
                updatedCount++;
            }
        }

        console.log(`Migration complete. Updated ${updatedCount} students.`);
        process.exit(0);
    } catch (err) {
        console.error("Migration failed:", err);
        process.exit(1);
    }
}

migrate();

File: SafeVote/backend/elections/management/commands/migrate_depts.py
from django.core.management.base import BaseCommand
from utils.database import get_students_collection

class Command(BaseCommand):
    help = "Migrate student departments based on roll number digits matching migrate_depts.js"

    def handle(self, *args, **options):
        self.stdout.write("Connected to MongoDB for department migration...")
        students_col = get_students_collection()
        students = list(students_col.find({}))
        self.stdout.write(f"Found {len(students)} students to check.")

        updated_count = 0
        for student in students:
            s_id = str(student.get("regNo", ""))
            dept_code = s_id[6:9] if len(s_id) >= 9 else ""

            dept = "OTHERS"
            if dept_code == "107":
                dept = "CYBER SECURITY"
            elif dept_code == "202":
                dept = "AIML"
            elif dept_code == "205":
                dept = "IOT"
            elif dept_code == "104":
                dept = "CSE"

            if student.get("department") != dept:
                students_col.update_one(
                    {"_id": student["_id"]},
                    {"$set": {"department": dept}}
                )
                updated_count += 1

        self.stdout.write(self.style.SUCCESS(f"Migration complete. Updated {updated_count} students."))
